### PR TITLE
Sci 1025 vcf qc

### PR DESCRIPTION
**Pull request by Arkarachai**
#dxapp.json
- [x] Upgrade to 14.04
- [x] Add help and extension to input spec.
- [x] In `details` field specify: `whatsNew`and `citation` (leave blank for citation.
- [x] In `details` consider specifying `upstreamAuthor` and `upstreamVersion` fields when applicable.
- [x] Use `regionalOptions` 
- [x] Set applicable timeout (Be relax) and restart policy in [`runSpec`](https://wiki.dnanexus.com/API-Specification-v1.0.0/IO-and-Run-Specifications).
- [x] Add subjob restartability when applicable.
- [x] `version` field, set proper [semantic versioning](https://semver.org/).
- [x] Setting `"openSource": true`. 
- [x] Run [dxapp linter](https://github.com/dnanexus/RoughER/blob/master/scripts/dxapp_lint.py) [[permalink](https://github.com/dnanexus/RoughER/blob/ecb3ea3f97870c020f93bf4723afc9f40a310e33/scripts/dxapp_lint.py)] on `dxapp.json`.
- [x] Use AssetDepend instead of execDepend
- [x] Add default instance type that works for a general use case.
- [x] add new input if applicable

#src/code.sh src/code.py
- [x] GATK tarball should be removed from the resource project (or make a new resource project), and the applet should require GATK tarball as a required input. This way, the users are enforced to provide their own tarball. This would avoid trouble from non-academic entities without GATK license to use your tarball.

# Readme.md
- [x] Improve Readme file to contain applet function, use case, input/output description, underline execution if it’s a wrapper.
- [ ] Read documentation for sanity check. (Unpublished app URL will be: https://platform.dnanexus.com/app/app-xxxx)

# Test
- [x] Basic testing using given input in Box or expected input
- [] Make a Pull Request to `master` in https://github.com/RitchieLab/dnanexus

# Production
- [ ] Co-test with Ritchie group
- [ ] Help Ritchie group to publish (Give dx command)
- [ ] Verify the versions that appear on the platform are the intended versions
- [ ] Set up demo workflow in public project.

### CODE REVIEW

File: sequencing/vcf_qc/dxapp.json [hunk 1]
```diff
@@ -102,62 +172,39 @@
...
-        "instanceType": "mem1_ssd2_x8"
-      }
-    },
     "distribution": "Ubuntu",
     "release": "14.04",
+    "restartableEntryPoints": "all", 
     "interpreter": "bash",
-    "file": "src/vcf_qc.sh"
+    "file": "src/vcf_qc.sh",
+    "timeoutPolicy": {"*": { "hours": 120 }}
```
Comment (nainathangaraj) [inline sequencing/vcf_qc/dxapp.json]: Maybe we can reduce this to 2 or 3 days instead of 5 days?

**Review by nainathangaraj — APPROVED** [review]
Looks good, Chai! The only thing that needs to be included is the naming convention of the inputs and the help field in the dxapp.json.

File: sequencing/vcf_qc/dxapp.json (hunk 1 above)
Comment (Arkarachai) [reply sequencing/vcf_qc/dxapp.json]: Thanks Naina.

**Comment by Arkarachai** [comment]
@slzarate Could you check the README of this applet? Thanks.

File: sequencing/vcf_qc/Readme.md [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
-This is the source code for an app that runs on the DNAnexus Platform.
-For more information about how to run or modify it, see
-https://wiki.dnanexus.com/.
-<!-- /dx-header -->
+This app uses [GATK ApplyRecalibration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_variantrecalibration_VariantRecalibrator.php) and [GATK VariantFiltration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php) to apply filters to VCF files.
 
+## What data are required for this app to run?
 
+This app requires:
+1. The array of VCF (variant calling format) files (`*.vcf.gz`) and their associated index files (`*.vcf.gz.tbi`). VCF is the standard format for variant calling.
```
Comment (slzarate) [inline sequencing/vcf_qc/Readme.md]: Do we have to define VCF here? Wouldn't the target audience of this app know this already?

File: sequencing/vcf_qc/Readme.md [hunk 3]
```diff
@@ -1,21 +1,19 @@
...
+2. The GATK 3 tar ball
+3. The reference genome in FASTA format (`*.fasta` or `*.fasta`), the Picard FASTA index (`*.dict`), and the satools FASTA index (`*.fasta.fai`)
```
Comment (slzarate) [inline sequencing/vcf_qc/Readme.md]: "satools" --> "samtools"

File: sequencing/vcf_qc/Readme.md [hunk 4]
```diff
@@ -1,21 +1,19 @@
...
-For more information about how to run or modify it, see
-https://wiki.dnanexus.com/.
-<!-- /dx-header -->
+This app uses [GATK ApplyRecalibration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_variantrecalibration_VariantRecalibrator.php) and [GATK VariantFiltration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php) to apply filters to VCF files.
 
+## What data are required for this app to run?
 
+This app requires:
+1. The array of VCF (variant calling format) files (`*.vcf.gz`) and their associated index files (`*.vcf.gz.tbi`). VCF is the standard format for variant calling.
+2. The GATK 3 tar ball
```
Comment (slzarate) [inline sequencing/vcf_qc/Readme.md]: "tar ball" --> "tarball"

File: sequencing/vcf_qc/Readme.md [hunk 5]
```diff
@@ -1,21 +1,19 @@
...
+3. The reference genome in FASTA format (`*.fasta` or `*.fasta`), the Picard FASTA index (`*.dict`), and the satools FASTA index (`*.fasta.fai`)
 
-<!--
-TODO: This app directory was automatically generated by dx-app-wizard;
-please edit this Readme.md file to include essential documentation about
-your app that would be helpful to users. (Also see the
-Readme.developer.md.) Once you're done, you can remove these TODO
-comments.
+The user can also provide the BED interval of exon target and other options for variant filtering.
```
Comment (slzarate) [inline sequencing/vcf_qc/Readme.md]: "Users can also provide a BED interval file containing exon targets **and other options for variant filtering**." (Would the BED file also contain other options for variant filtering?)

File: sequencing/vcf_qc/dxapp.json [hunk 6]
```diff
@@ -3,35 +3,108 @@
...
   "dxapi": "1.0.0",
-  "version": "0.0.2",
+  "version": "1.0.0",
   "categories": [
     "Variation Calling"
   ],
   "inputSpec": [
     {
-      "name": "vcf_fn",
+      "name": "fn_vcfgz",
```
Comment (slzarate) [inline sequencing/vcf_qc/dxapp.json]: Is this our standard name for VCF files?

Comment (Arkarachai) [reply sequencing/vcf_qc/dxapp.json]: Not quite. We will discuss this issue on July 11.

**Comment by Arkarachai** [comment]
@slzarate @nainathangaraj Thanks for all the comments and correction!

File: sequencing/vcf_qc/Readme.md [hunk 7]
```diff
@@ -1,21 +1,19 @@
-<!-- dx-header -->
 # VCF QC (DNAnexus Platform App)
 
-Calls [GATK ApplyRecalibration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_variantrecalibration_VariantRecalibrator.php) and [GATK VariantFiltration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php) to apply filters to VCF files.
+## What does this applet do?
```
Comment (yihchii) [inline sequencing/vcf_qc/Readme.md:3]: "What does this applet do?" --> "What does this app do?"

File: sequencing/vcf_qc/Readme.md [hunk 8]
```diff
@@ -1,21 +1,19 @@
...
 
-This is the source code for an app that runs on the DNAnexus Platform.
-For more information about how to run or modify it, see
-https://wiki.dnanexus.com/.
-<!-- /dx-header -->
+This app uses [GATK ApplyRecalibration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_variantrecalibration_VariantRecalibrator.php) and [GATK VariantFiltration](https://www.broadinstitute.org/gatk/guide/tooldocs/org_broadinstitute_gatk_tools_walkers_filters_VariantFiltration.php) to apply filters to VCF files.
```
Comment (yihchii) [inline sequencing/vcf_qc/Readme.md]: "GATK ApplyRecalibration" -> "GATK3 ApplyRecalibration"
"GATK VariantFiltration" -> "GATK3 VariantFiltration"

File: sequencing/vcf_qc/dxapp.json (hunk 6 above)
Comment (yihchii) [reply sequencing/vcf_qc/dxapp.json]: "fn_vcfgz" -> "variants_vcfgzs"

File: sequencing/vcf_qc/dxapp.json [hunk 9]
```diff
@@ -3,35 +3,108 @@
...
       "label": "VCF File",
       "class": "array:file",
       "optional": false,
-      "patterns": [
-        "*.vcf.gz"
-      ]
+      "patterns": ["*.vcf.gz"],
+      "help": "Array of VCF.gz files."
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json:17]: "Array of VCF.gz files." -> "Array of VCF files."
Also need to update in the script.

File: sequencing/vcf_qc/dxapp.json [hunk 10]
```diff
@@ -3,35 +3,108 @@
...
     },
     {
-      "name": "vcfidx_fn",
+      "name": "fn_vcfidx",
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "fn_vcfidx" -> "variants_vcfgztbis"
Also need to update in the script.

File: sequencing/vcf_qc/dxapp.json [hunk 11]
```diff
@@ -3,35 +3,108 @@
...
       "label": "VCF Index File",
       "class": "array:file",
       "optional": false,
-      "patterns": [
-        "*.tbi"
+      "patterns": ["*.tbi", "*.vcf.gz.tbi"],
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "patterns": ["\*.tbi", "\*.vcf.gz.tbi"], -> "patterns": ["\*.tbi"],

File: sequencing/vcf_qc/dxapp.json [hunk 12]
```diff
@@ -3,35 +3,108 @@
...
+      "help": "A FASTA file of reference genome.",
+      "class": "file",
+      "patterns": ["*.fasta","*.fa","*.fasta.gz","*.fa.gz"],
+      "optional": false,
+      "suggestions": [
+        {
+          "project": "project-FGqF5B00ByvPXG4G87yQy23G",
+          "path": "/GATK/resources",
+          "region": "aws:us-east-1",
+          "name": "dx_ritchie_collaboration: AWS US-east"
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json:47]: Not sure about the name used here. We (Ritchie lab and DNAnexus) may need to update to a generic name before paper submission and app publishing.

File: sequencing/vcf_qc/dxapp.json [hunk 13]
```diff
@@ -3,35 +3,108 @@
...
+        },
+        {
+          "project": "project-FGvp3009BJq40pkZ683v3754",
+          "path": "/GATK/resources",
+          "region": "aws:westus",
+          "name": "dx_ritchie_collaboration: Azure West US"
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json:53]: Same as above. Not sure about the name used here. We (Ritchie lab and DNAnexus) may need to update to a generic name before paper submission and app publishing.

File: sequencing/vcf_qc/dxapp.json [hunk 14]
```diff
@@ -3,35 +3,108 @@
...
+      "help": "A dict file of FASTA reference genome generated by Picard.",
+      "class": "file",
+      "patterns": ["*.dict"],
+      "optional": false,
+      "suggestions": [
+        {
+          "project": "project-FGqF5B00ByvPXG4G87yQy23G",
+          "path": "/GATK/resources",
+          "region": "aws:us-east-1",
+          "name": "dx_ritchie_collaboration: AWS US-east"
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json:69]: Same as above. Not sure about the name used here. We (Ritchie lab and DNAnexus) may need to update to a generic name before paper submission and app publishing.

File: sequencing/vcf_qc/dxapp.json [hunk 15]
```diff
@@ -3,35 +3,108 @@
...
+        },
+        {
+          "project": "project-FGvp3009BJq40pkZ683v3754",
+          "path": "/GATK/resources",
+          "region": "aws:westus",
+          "name": "dx_ritchie_collaboration: Azure West US"
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json:75]: Same as above. Not sure about the name used here. We (Ritchie lab and DNAnexus) may need to update to a generic name before paper submission and app publishing.

File: sequencing/vcf_qc/dxapp.json [hunk 16]
```diff
@@ -3,35 +3,108 @@
...
+        }
+      ]
+    },
+    {
+      "name": "genome_fasta_fai",
+      "label": "Reference genome fai",
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "Reference genome fai" - > "Reference genome index"

File: sequencing/vcf_qc/dxapp.json [hunk 17]
```diff
@@ -3,35 +3,108 @@
...
+      "help": "A FASTA fai index of reference genome generated by samtools.",
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "A FASTA fai index of reference genome generated by samtools." -> "A FASTA FAI index of reference genome generated by samtools."

File: sequencing/vcf_qc/dxapp.json [hunk 18]
```diff
@@ -102,62 +175,39 @@
...
-      "choices": ["rgc_b38","gatk_b38","b37_decoy","hg38chr.fa"],
-      "default": "rgc_b38",
-      "optional": false
     }
   ],
   "outputSpec": [
     {
-      "name": "vcf_out",
-      "label": "Filtered VCF file",
+      "name": "out_vcfgz",
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "out_vcfgz" -> "out_variants_vcfgzs"
Also need to update in the script accordingly.

File: sequencing/vcf_qc/dxapp.json [hunk 19]
```diff
@@ -102,62 +175,39 @@
...
       "class": "array:file",
-      "patterns": [
-        "*.vcf.gz"
-      ]
+      "patterns": ["*.vcf.gz"]
     },
     {
-      "name": "vcfidx_out",
-      "label": "Filtered VCF Index file",
+      "name": "out_vcfidx",
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "out_vcfidx" -> "out_variants_vcfgztbis"
Also need to update in the script accordingly.

File: sequencing/vcf_qc/dxapp.json [hunk 20]
```diff
@@ -102,62 +175,39 @@
...
-  "access": {
-    "network": [
-      "*"
-    ],
-     "allProjects": "VIEW"
+  "regionalOptions": {
+    "aws:us-east-1": {"systemRequirements": {"*": {"instanceType": "mem1_ssd2_x8"}},
+                    "assetDepends": [{"id": "record-FGpf7zQ0kfbZxj6q4ZzQ2BPG"}]},
+    "azure:westus": {"systemRequirements": {"*": {"instanceType": "azure:mem2_ssd1_x16"}},
+                    "assetDepends": [{"id": "record-FGpfZ1j95qzXV43bP5y5JKgV"}]}
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: Please update record IDs to:
aws:us-east-1: record-FJ3zf8j08zVZFzqYG88Pbk7z
azure:westus: record-FJ4058j9Vf9ZBBVB2Q7b3VQz

**Review by yihchii — COMMENTED** [review]
LGTM. 
Please 
1. Update dxapp.json and script to follow input/output naming nomenclature.
2. Update asset record IDs to use the up-to-date ones.

File: sequencing/vcf_qc/dxapp.json [hunk 21]
```diff
@@ -102,62 +175,39 @@
...
   },
-  "developers": ["user-tnperson", "user-auv13"],
   "openSource" : true,
-  "authorizedUsers" : ["org-geisinger_affiliate", "org-geisinger", "org-support", "user-codushlaine", "user-semanti.mukherjee", "user-alexander.li", "user-shannon.bruse", "user-jan.freudenberg", "user-crisvh"]
+  "details": {
+    "upstreamUrl": "https://software.broadinstitute.org/gatk/",
+    "whatsNew": ""
```
Comment (yihchii) [inline sequencing/vcf_qc/dxapp.json]: "whatsNew": "* 1.0.0 First version of the VCF QC app."

**Comment by msimbirsky** [comment]
@yihhwang @slzarate Changes implemented

**Pull request merged**